src/ppx/xr_composition_layers.cpp

```diff
@@ -0,0 +1,61 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#if defined(PPX_BUILD_XR)
+#include "ppx/xr_composition_layers.h"
+
+#include <memory>
+#include <optional>
+#include <utility>
+#include <vector>
+
+namespace ppx {
+
+XrProjectionLayer::XrProjectionLayer()
+{
+    layer().views = mViews.data();
+}
+
+void XrProjectionLayer::AddView(XrCompositionLayerProjectionView view)
+{
+    mDepthInfos.push_back(std::nullopt);
+    mViews.emplace_back(view);
+    FixViewReferences();
+}
+
+void XrProjectionLayer::AddView(XrCompositionLayerProjectionView view, XrCompositionLayerDepthInfoKHR depthInfo)
+{
+    mDepthInfos.emplace_back(std::make_optional(depthInfo));
+    mViews.emplace_back(view);
+    FixViewReferences();
+}
+
+void XrProjectionLayer::FixViewReferences()
+{
+    for (size_t i = 0; i < mViews.size(); i++) {
+        if (mDepthInfos[i].has_value()) {
```

Comment:
This seems a bit clunky, how about using a list for depth infos, vector of unique_ptr, or a fixed size array (since I think the # of the objects is relatively small)?

Reply:
Done. Used `std::vector<std::unqiue_ptr<Struct>>` to keep it generic across any potential usecase.